// react-native-esp32-idf/lib/module/index.js
import { NativeModules, NativeEventEmitter, AppState } from 'react-native';
import { useEffect, useState, useRef } from 'react';
const {
  RNEsp32Idf
} = NativeModules;
const eventEmitter = new NativeEventEmitter(RNEsp32Idf);
export default RNEsp32Idf;
export function useProvisioning({
  devicePrefix,
  pop = null,
  message
}) {
  console.log('Invoke func useProvisioning');
  const msg = useRef(message);
  const [bleDevices, setBleDevices] = useState([]);
  const [wifiAPs, setWifiAPs] = useState([]);
  const [loading, setLoading] = useState(true);
  const [status, setStatus] = useState(message.scanBle);
  const isConnecting = useRef(false);
  const [currentStep, setCurrentStep] = useState(1);
  const [provSent, setProvSent] = useState(initStep(message.sendingWifiCredential, true));
  const [provApplied, setProvApplied] = useState(initStep(message.confirmWifiConnection));
  const [provFinal, setProvFinal] = useState(initStep(''));
  const currentWifi = useRef();
  const currentDevice = useRef();

  function connectDevice(bleDevice) {
    if (isConnecting.current) return;
    isConnecting.current = true;
    RNEsp32Idf.stopBleScan();
    console.log('Connect to device:', bleDevice, pop);
    setStatus(message.connectDevice);
    currentDevice.current = bleDevice;
    RNEsp32Idf.connectDevice(bleDevice.serviceUuid, pop);
  }

  function configWifi(wifi) {
    console.log('Selected Wifi:', wifi);
    currentWifi.current = wifi;
    setCurrentStep(!wifi.ssid || wifi.auth > 0 ? 3 : 4);

    if (!wifi.auth) {
      doProvisioning(wifi);
    }
  }

  useEffect(() => {
    eventEmitter.addListener('scanWifi', event => {
      console.log('Event scanWifi', event);
      setLoading(false);
      setStatus('');

      if (event.wifiList) {
        setWifiAPs(event.wifiList);
      } else if (event.message) {
        setStatus(event.message);
      }
    });
    return function () {
      console.log('Cleanup the resource');
      RNEsp32Idf.stopBleScan();
      RNEsp32Idf.disconnectDevice();
      eventEmitter.removeAllListeners('scanWifi');
    };
  }, []);
  useEffect(() => {
    console.log('Added AppSateChanged listener');

    async function _handleAppStateChange(nextAppState) {
      console.log(`AppSateChanged: ${nextAppState}`);

      if (nextAppState === 'active') {
        console.log('Start checkPermissions');
        const result = await RNEsp32Idf.checkPermissions();
        console.log(`checkPermissions result: ${result}`);

        if (result && !isConnecting.current && currentStep === 1) {
          console.log('Start BleScan');
          RNEsp32Idf.startBleScan(devicePrefix);
        }
      }
    }

    AppState.addEventListener('change', _handleAppStateChange);
    eventEmitter.addListener('permission', event => {
      console.log('Event permission', event);

      if (event.status) {
        if (currentStep === 1 && !isConnecting.current) {
          RNEsp32Idf.startBleScan(devicePrefix);
          setBleDevices([]);
        } else if (currentStep === 2) {
          RNEsp32Idf.startWifiScan();
        }
      } else {
        setStatus(event.type === 1 ? msg.current.enableBluetooth : msg.current.enableLocation);
      }
    });
    return function () {
      console.log('Removed AppSateChanged listener');
      eventEmitter.removeAllListeners('permission');
      AppState.removeEventListener('change', _handleAppStateChange);
    };
  }, [currentStep, devicePrefix]);
  useEffect(() => {
    console.log('Added listeners');
    eventEmitter.addListener('scanBle', event => {
      console.log('Event scanBle', event);

      if (event instanceof Array) {
        setBleDevices(event);
      } else if (event.deviceName) {
        setBleDevices(prev => prev.some(it => it.serviceUuid === event.serviceUuid) ? prev : prev.concat(event));
      } else if (event.status === 0) {
        setLoading(false);
        setStatus(msg.current.scanBleFailed);
      } else {
        setLoading(false);
        if (!isConnecting.current) setStatus('');
      }
    });
    eventEmitter.addListener('connection', event => {
      isConnecting.current = false;
      console.log('Event connection', event);

      switch (event.status) {
        case 1:
          //connected
          RNEsp32Idf.startWifiScan();
          setCurrentStep(2);
          setLoading(true);
          setStatus(msg.current.scanWifi);
          break;

        case 2:
          //failed
          setStatus(msg.current.connectFailed);
          break;

        case 3:
          //disconnected
          setStatus(msg.current.disconnected);
          break;
      }
    });
    eventEmitter.addListener('provisioning', event => {
      console.log('Event provisioning', event);

      switch (event.status) {
        case 0:
        case 2:
          setProvSent(doneStep(event.message, true));
          setProvFinal(doneStep(msg.current.initSessionError, true));
          break;

        case 3:
          setProvSent(prev => ({ ...prev,
            done: true
          }));
          setProvApplied(prev => ({ ...prev,
            progress: true
          }));
          break;

        case 5:
          setProvApplied(prev => ({ ...prev,
            done: true
          }));
          setProvFinal(doneStep(msg.current.completed));
          break;

        default:
          setProvApplied(doneStep(event.message, false));
          setProvFinal(doneStep(msg.current.applyError, true));
      }
    });
    return function () {
      console.log('Removed listeners');
      eventEmitter.removeAllListeners('scanBle');
      eventEmitter.removeAllListeners('connection');
      eventEmitter.removeAllListeners('provisioning');
    };
  }, [devicePrefix]);
  return {
    bleDevices,
    wifiAPs,
    loading,
    status,
    currentStep,
    currentWifi,
    currentDevice,
    results: [provSent, provApplied, provFinal],
    setCurrentStep,
    connectDevice,
    configWifi,
    doProvisioning
  };
}

function doProvisioning(_wifi) {
  RNEsp32Idf.doProvisioning(_wifi.ssid, 'password' in _wifi ? _wifi.password : '');
}

function initStep(message, progress = false) {
  return {
    done: false,
    progress,
    failed: false,
    message
  };
}

function doneStep(message, failed = false) {
  return {
    progress: false,
    done: true,
    failed,
    message
  };
}
